File: mobile/ios/Actnet/Sources/Services/DevServerActnetService.swift
import Foundation

/// Service that hits a real locally-running homeserver via the Rust AppCore.
struct DevServerActnetService: ActnetService {
    static let defaultServerUrl = "http://localhost:3000"

    func createAccount(serverUrl: String, dbPath: String, dbKey: String, prfOutput: Data, displayName: String, inviteToken: String?) throws -> any AppCoreProtocol {
        try AppCore.createAccount(serverUrl: serverUrl, dbPath: dbPath, dbKey: dbKey, prfOutput: prfOutput, displayName: displayName, inviteToken: inviteToken)
    }

    func login(dbPath: String, dbKey: String) throws -> any AppCoreProtocol {
        try AppCore.login(dbPath: dbPath, dbKey: dbKey)
    }

    func prepareAccount(serverUrl: String, prfOutput: Data) throws -> any PreparedAccountProtocol {
        try PreparedAccount(serverUrl: serverUrl, prfOutput: prfOutput)
    }

    func finalizeAccount(prepared: any PreparedAccountProtocol, dbPath: String, dbKey: String, displayName: String, inviteToken: String?) throws -> any AppCoreProtocol {
        guard let concrete = prepared as? PreparedAccount else {
            throw ActnetServiceError.preparedAccountTypeMismatch
        }
        return try AppCore.finalizeAccount(prepared: concrete, dbPath: dbPath, dbKey: dbKey, displayName: displayName, inviteToken: inviteToken)
    }

    func recoverFromBlob(serverUrl: String, did: String, prfOutput: Data, dbPath: String, dbKey: String, displayName: String) throws -> any AppCoreProtocol {
        try AppCore.recoverFromBlob(serverUrl: serverUrl, did: did, prfOutput: prfOutput, dbPath: dbPath, dbKey: dbKey, displayName: displayName)
    }

    func makeDeviceLink() -> any DeviceLinkProtocol {
        LiveDeviceLink()
    }
}

/// Live device-link handle wrapping the UniFFI `DeviceLinkNew` object.
final class LiveDeviceLink: DeviceLinkProtocol, @unchecked Sendable {
    private let inner = DeviceLinkNew()

    func createPairing(mailboxServer: String?) throws -> String {
        try inner.createPairing(mailboxServer: mailboxServer)
    }

    func acceptPairing(code: String) throws {
        try inner.acceptPairing(code: code)
    }

    func awaitLinkStep(dbPath: String, dbKey: String) throws -> (any AppCoreProtocol)? {
        try inner.awaitLinkStep(dbPath: dbPath, dbKey: dbKey)
    }
}

enum ActnetServiceError: LocalizedError {
    case preparedAccountTypeMismatch

    var errorDescription: String? {
        switch self {
        case .preparedAccountTypeMismatch:
            return "PreparedAccount instance does not match the active ActnetService."
        }
    }
}

File: mobile/android/app/src/main/kotlin/Services/DevServerActnetService.kt
package net.theavalanche.app

import uniffi.app_core.AppErrorFfi
import uniffi.app_core.PreparedAccount

/**
 * Live [ActnetService] backed by a real locally-running homeserver via the
 * Rust AppCore UniFFI bindings.
 *
 * Mirrors iOS `DevServerActnetService` — a concrete implementation of the
 * `ActnetService` protocol that forwards all account-lifecycle calls to the
 * UniFFI-generated [AppCore] static constructors.
 *
 * All methods are synchronous and BLOCKING. Callers (ViewModel layer) must
 * dispatch onto [kotlinx.coroutines.Dispatchers.IO] before invoking.
 */
class DevServerActnetService : ActnetService {

    companion object {
        const val DEFAULT_SERVER_URL = "http://localhost:3000"
    }

    @Throws(AppErrorFfi::class)
    override fun createAccount(
        serverUrl: String,
        dbPath: String,
        dbKey: String,
        prfOutput: ByteArray,
        displayName: String,
        inviteToken: String?,
    ): AppCoreProtocol =
        LiveActnetService.createAccount(serverUrl, dbPath, dbKey, prfOutput, displayName, inviteToken)

    @Throws(AppErrorFfi::class)
    override fun login(dbPath: String, dbKey: String): AppCoreProtocol =
        LiveActnetService.login(dbPath, dbKey)

    @Throws(AppErrorFfi::class)
    override fun prepareAccount(serverUrl: String, prfOutput: ByteArray): PreparedAccount =
        LiveActnetService.prepareAccount(serverUrl, prfOutput)

    @Throws(AppErrorFfi::class)
    override fun finalizeAccount(
        prepared: PreparedAccount,
        dbPath: String,
        dbKey: String,
        displayName: String,
        inviteToken: String?,
    ): AppCoreProtocol =
        LiveActnetService.finalizeAccount(prepared, dbPath, dbKey, displayName, inviteToken)

    @Throws(AppErrorFfi::class)
    override fun recoverFromBlob(
        serverUrl: String,
        did: String,
        prfOutput: ByteArray,
        dbPath: String,
        dbKey: String,
        displayName: String,
    ): AppCoreProtocol =
        LiveActnetService.recoverFromBlob(serverUrl, did, prfOutput, dbPath, dbKey, displayName)

    override fun makeDeviceLink(): DeviceLink = LiveActnetService.makeDeviceLink()
}

/**
 * Mirrors iOS `ActnetServiceError` — errors that can arise from the
 * [ActnetService] layer itself (distinct from [AppErrorFfi] thrown by the Rust core).
 */
sealed class ActnetServiceError(message: String) : Exception(message) {
    /** Thrown by [DevServerActnetService.finalizeAccount] when the [PreparedAccount] argument
     *  was not produced by the same service implementation. */
    object PreparedAccountTypeMismatch : ActnetServiceError(
        "PreparedAccount instance does not match the active ActnetService."
    )
}
